File: results/2026-06-30_191904/tasks/13-dependency-license-checker/default-sonnet5-1m-medium/generated-code/manifest_parser.py
"""
Parsers for dependency manifests (package.json, requirements.txt).

Each parser returns a plain dict of {dependency_name: version_spec}.
Version specs are kept as-is from the manifest (e.g. "^1.3.0", "==2.31.0"
is normalized to "2.31.0") so downstream consumers decide how to interpret
range operators.
"""
import json
import os
import re


class ManifestParseError(Exception):
    """Raised when a manifest file cannot be read or parsed."""


def _read_file(path):
    if not os.path.isfile(path):
        raise ManifestParseError(f"Manifest file not found: {path}")
    with open(path, "r", encoding="utf-8") as fh:
        return fh.read()


def parse_package_json(path):
    """Parse an npm package.json, merging dependencies + devDependencies."""
    raw = _read_file(path)
    try:
        data = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise ManifestParseError(f"Invalid JSON in {path}: {exc}") from exc

    deps = {}
    deps.update(data.get("dependencies", {}) or {})
    deps.update(data.get("devDependencies", {}) or {})
    return deps


# Matches "name==1.2.3", "name>=1.2.3", "name~=1.2.3", or bare "name"
_REQ_LINE_RE = re.compile(
    r"^\s*([A-Za-z0-9._-]+)\s*(==|>=|<=|~=|!=|>|<)?\s*([A-Za-z0-9._-]*)\s*$"
)


def parse_requirements_txt(path):
    """Parse a pip requirements.txt into {name: version}. Unpinned -> '*'."""
    raw = _read_file(path)
    deps = {}
    for line in raw.splitlines():
        stripped = line.strip()
        if not stripped or stripped.startswith("#") or stripped.startswith("-"):
            continue
        match = _REQ_LINE_RE.match(stripped)
        if not match:
            continue
        name, _op, version = match.groups()
        deps[name] = version if version else "*"
    return deps


_PARSERS_BY_FILENAME = {
    "package.json": parse_package_json,
    "requirements.txt": parse_requirements_txt,
}


def parse_manifest(path):
    """Dispatch to the right parser based on the manifest's filename."""
    basename = os.path.basename(path)
    parser = _PARSERS_BY_FILENAME.get(basename)
    if parser is None:
        raise ManifestParseError(
            f"Unsupported manifest type: {basename!r}. "
            f"Supported types: {sorted(_PARSERS_BY_FILENAME)}"
        )
    return parser(path)

File: results/2026-06-30_191904/tasks/13-dependency-license-checker/typescript-bun-sonnet5/generated-code/src/manifestParser.ts
import { existsSync, readFileSync } from "node:fs";
import { basename } from "node:path";
import type { Dependency } from "./types";

/**
 * Parses a dependency manifest file and returns the flat list of
 * declared dependencies. Supports npm's package.json (dependencies +
 * devDependencies) and Python's requirements.txt.
 *
 * The manifest format is chosen by filename, not extension, since
 * requirements.txt has no extension of its own to key off.
 */
export function parseManifest(filePath: string): Dependency[] {
  if (!existsSync(filePath)) {
    throw new Error(`Manifest file not found: ${filePath}`);
  }

  const contents = readFileSync(filePath, "utf-8");
  const name = basename(filePath);

  if (name === "package.json") {
    return parsePackageJson(filePath, contents);
  }
  if (name === "requirements.txt") {
    return parseRequirementsTxt(contents);
  }

  throw new Error(
    `Unsupported manifest type for "${filePath}". Expected package.json or requirements.txt.`,
  );
}

function parsePackageJson(filePath: string, contents: string): Dependency[] {
  let parsed: unknown;
  try {
    parsed = JSON.parse(contents);
  } catch (cause) {
    throw new Error(`Invalid JSON in manifest file: ${filePath}`, { cause });
  }

  if (typeof parsed !== "object" || parsed === null) {
    throw new Error(`Invalid JSON in manifest file: ${filePath}`);
  }

  const pkg = parsed as Record<string, unknown>;
  const deps: Dependency[] = [];

  for (const section of ["dependencies", "devDependencies"] as const) {
    const sectionValue = pkg[section];
    if (!sectionValue || typeof sectionValue !== "object") continue;

    for (const [depName, rawVersion] of Object.entries(
      sectionValue as Record<string, unknown>,
    )) {
      deps.push({
        name: depName,
        version: cleanSemverRange(String(rawVersion)),
      });
    }
  }

  return deps;
}

/** Strips leading range operators (^ ~ >= <= > < =) so we're left with a bare version. */
function cleanSemverRange(rawVersion: string): string {
  const trimmed = rawVersion.trim();
  if (trimmed === "*" || trimmed === "") return "*";
  return trimmed.replace(/^[\^~]|^(>=|<=|==|>|<|=)/, "").trim();
}

const REQUIREMENT_LINE = /^([A-Za-z0-9._-]+)\s*(==|>=|<=|~=|!=|>|<)?\s*(.*)$/;

function parseRequirementsTxt(contents: string): Dependency[] {
  const deps: Dependency[] = [];

  for (const rawLine of contents.split(/\r?\n/)) {
    const line = rawLine.trim();
    if (line === "" || line.startsWith("#") || line.startsWith("-")) continue;

    const match = REQUIREMENT_LINE.exec(line);
    if (!match) continue;

    const [, depName, , versionPart] = match;
    const version = versionPart && versionPart.trim() !== "" ? versionPart.trim() : "*";
    deps.push({ name: depName as string, version });
  }

  return deps;
}
